File: java/CavanJava/src/com/cavan/java/CavanPeakValleyValue.java
package com.cavan.java;

public class CavanPeakValleyValue {

	public static final int TYPE_FALLING = 0;
	public static final int TYPE_RISING = 1;

	protected int mType;
	protected long mTime;
	protected double mPeakValue;
	protected double mValleyValue;

	public CavanPeakValleyValue(double peak, double valley, int type, long time) {
		super();

		mType = type;
		mTime = time;
		mPeakValue = peak;
		mValleyValue = valley;
	}

	public CavanPeakValleyValue(double peak, double valley, int type) {
		this(peak, valley, type, System.currentTimeMillis());
	}

	public CavanPeakValleyValue(double peak, double valley) {
		this(peak, valley, TYPE_FALLING);
	}

	public CavanPeakValleyValue(double value) {
		this(value, value);
	}

	public double getPeakValue() {
		return mPeakValue;
	}

	public void setPeakValue(double peak) {
		mPeakValue = peak;
	}

	public double getValleyValue() {
		return mValleyValue;
	}

	public void setValleyValue(double valley) {
		mValleyValue = valley;
	}

	public int getType() {
		return mType;
	}

	public void setType(int type) {
		mType = type;
	}

	public long getTime() {
		return mTime;
	}

	public long getTimeEarly(long time) {
		return time - mTime;
	}

	public long getTimeEarly(CavanPeakValleyValue value) {
		return getTimeEarly(value.getTime());
	}

	public long getTimeLate(long time) {
		return mTime - time;
	}

	public long getTimeLate(CavanPeakValleyValue value) {
		return getTimeLate(value.getTime());
	}

	public long getTimeDiff(long time) {
		if (mTime > time) {
			return mTime - time;
		} else {
			return time - mTime;
		}
	}

	public long getTimeDiff(CavanPeakValleyValue value) {
		return getTimeDiff(value.getTime());
	}

	public boolean isFalling() {
		return mType == TYPE_FALLING;
	}

	public boolean isRising() {
		return mType == TYPE_RISING;
	}

	public double getDiff() {
		return mPeakValue - mValleyValue;
	}

	public void extend(CavanPeakValleyValue value) {
		if (value.getPeakValue() > mPeakValue) {
			mPeakValue = value.getPeakValue();
		}

		if (value.getValleyValue() < mValleyValue) {
			mValleyValue = value.getValleyValue();
		}
	}
}

File: xcode/JwaooToy/JwaooToy/CavanPeakValleyValue.m
//
//  CavanPeakValleyValue.m
//  JwaooToy
//
//  Created by 曹福昂 on 16/7/8.
//  Copyright © 2016年 曹福昂. All rights reserved.
//

#import "CavanPeakValleyValue.h"

@implementation CavanPeakValleyValue

@synthesize type = mType;
@synthesize time = mTime;
@synthesize peak = mPeakValue;
@synthesize valley = mValleyValue;

- (CavanPeakValleyValue *)initWithPeak:(double)peak
                            withValley:(double)valley
                              withType:(CavanPeakValleyValueType)type
                              withTime:(NSTimeInterval)time {
    if (self = [super init]) {
        mType = type;
        mTime = time;
        mPeakValue = peak;
        mValleyValue = valley;
    }

    return self;
}

- (CavanPeakValleyValue *)initWithPeak:(double)peak
                            withValley:(double)valley
                              withType:(CavanPeakValleyValueType)type {
    return [self initWithPeak:peak withValley:valley withType:type withTime:[NSDate timeIntervalSinceReferenceDate]];
}

- (CavanPeakValleyValue *)initWithValue:(double)value {
    return [self initWithPeak:value withValley:value withType:CavanPeakValleyValueFalling];
}

- (CavanPeakValleyValue *)copy {
    return [[self.class alloc] initWithPeak:mPeakValue withValley:mValleyValue withType:mType];
}

- (NSString *)description {
    if ([self isRising]) {
        return [NSString stringWithFormat:@"Rising: [%f, %f]", mValleyValue, mPeakValue];
    } else {
        return [NSString stringWithFormat:@"Falling: [%f, %f]", mPeakValue, mValleyValue];
    }
}

- (double)getDiff {
    return mPeakValue - mPeakValue;
}

- (BOOL)isFalling {
    return mType == CavanPeakValleyValueFalling;
}

- (BOOL)isRising {
    return mType == CavanPeakValleyValueRising;
}

- (NSTimeInterval)timeIntervalSinceTime:(NSTimeInterval)time {
    return mTime - time;
}

- (NSTimeInterval)timeIntervalSince:(CavanPeakValleyValue *)value {
    return [self timeIntervalSinceTime:value.time];
}

- (NSTimeInterval)timeIntervalSinceDate:(NSDate *)date {
    return [self timeIntervalSinceTime:[date timeIntervalSinceReferenceDate]];
}

@end
